cEP-0026.md

```diff
@@ -0,0 +1,189 @@
+
+| Metadata |                                           |
+| ---------| ----------------------------------------- |
+| cEP      | 26                                        |
+| Version  | 1.0                                       |
+| Title    | Optimize caching for the NextGen-Core     |
+| Authors  | Palash Nigam <mailto:npalash25@gmail.com> |
+| Status   | Implementation Due                        |
+| Type     | Feature                                   |
+
+## Abstract
+
+This document describes how a caching mechanism (an enhanced version of the one
+being used by the old core) will be implemented and integrated with the
+NextGen-Core as a part of this
+[GSoC project](https://summerofcode.withgoogle.com/projects/#6434190552203264).
+
+## Introduction
+
+Since the NextGen-Core aims to provide faster speed and better performance for
+coala-runs on large codebases it needs caching to achieve this kind of
+performance.
+
+The current caching used by the old core can be improved upon and integrated
+with the NextGen-Core by adding a few more enhancements to the current caching
+mechanism like `FileFactory` class to load files, `Directory` class
+to interface with the directories, ignoring unmodified directories during
+successive runs and adding cache control flags for better usability and
+providing the user various options to perform caching according to a project's
+needs.
+
+## Implementation
+
+The implementation will be divided into four parts
+
+- Prototype for `FileFactory` class implementation.
+- Prototype for `Directory` class implementation.
+- Ignore directories functionality.
+- Cache control flags.
+
+### Prototype of FileFactory class implementation
+
+This class will be implemented according to the Factory design pattern and will
+act as a proxy to dealing with files. This class will have various methods to
+provide the file contents in different forms for e.g. string, binary
+format, etc. The biggest benefit that we will get out of this class would be
+that we can replace the actual file contents in our file-dict with these
+objects which will itself be a performance improvement.
+
+The following is the prototype for `FileFactory` class.
+
+```python
+import linecache
+import os
+import pickle
+
+from coalib.core.PersistentHash import persistent_hash
+
+
+class FileFactory:
+    def __init__(self, filename, parent):
+        self.filename = filename # absolute file path
+        self.parent = os.path.abspath(os.path.join(p, os.pardir)) # parent dir
+        self._content = open(self.filename)
+
+    def open(filename, newline=None):
+        return FileFactory(filename, newline=newline)
+
+    def timestamp(self):
+        # get the last modified timestamp of the file
+        return os.path.gettime(self.path)
+
+    def file_as_string(self, encoding=None):
+        return "".join(self._content)
+
+    def file_as_bytes(self):
+        # return the pickled version of the self._content
+        pickle.dumps(self.filename)
```

Comment:
nah just open the file in `binary` mode, and do `file.read()`. For `file_as_string` you can probably convert bytes via `file_as_bytes.decode()`

Reply:
If you want to normalize newlines additionally, you can maybe use `StringIO` and its `newline` parameter :)

Reply:
> For file_as_string you can probably convert bytes via file_as_bytes.decode()

What if we don't need to call `file_as_bytes` for certain factories in that case the current `file_as_string` method doesn't depend on the `file_as_bytes` method to generate its own value and can use reuse `self._content` from the `__init__` method which will be generated for every object that way we don't need to call the `file_as_bytes` method unnecessarily for generating the return value of another method (and end up not using its own return value, the file contents in binary) and if we have to use the binary form of the file content for certain objects we are gonna have  to call `file_as_bytes` again it leads to duplication. Should I keep it the way is is?

I can still change it if you want. :smile:

Reply:
See https://github.com/coala/cEPs/pull/146/#discussion_r189426373